File: python/23.05/74.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for i in range(len(matrix)):
            if matrix[i][0] > target:
                break
        else:
            i += 1
        i -= 1
        if i == -1:
            return False
        
        for j in matrix[i]:
            if j == target:
                return True
        
        return False

File: typescript/23.08/74.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    const bisect = (arr: number[], tar: number) => {
        let [low, mid, top] = [0, 0, arr.length]
        while (low < top) {
            mid = Math.floor((low+top)/2)
            if (arr[mid] == tar) {
                return mid + 1
            }
            if (arr[mid] < tar) {
                low = mid + 1
            } else {
                top = mid
            }
        }
        return low
    }

    const idx = bisect(matrix.filter(x => x).map(x => x[0]), target)
    if (idx <= 0) return false
    return matrix[idx-1][bisect(matrix[idx-1], target) - 1] === target ? true : false
};
